### PR TITLE
vimPlugins.fzf-vim: automatically add fzfWrapper

### PR DESCRIPTION
**Pull request by timokau**
###### Motivation for this change

fzf-vim doesn't work without "regular" fzf. Currently that makes it necessary to manually add both `fzf-vim` and `fzfWrapper`.

I don't think the `symlinkJoin` approach is ideal, but it is the best I could come up with (not being very familiar with the vim plugin infrastructure). Just adding `"fzfWrapper"` to the dependencies of `fzf-vim` didn't work.

CC @andsild

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by Ma27** [comment]
> Just adding `"fzfWrapper"` to the dependencies of `fzf-vim` didn't work.

Can you recheck, please? I just tried this with the following expression on master (dee617cb9775ac9b10745ba412a3cb6ee7cdc75a):

``` nix
with import ./. { };

vim_configurable.customize {
  name = "vim";
  vimrcConfig.vam = {
    knownPlugins = pkgs.vimPlugins;
    pluginDictionaries = [
      { name = "fzfWrapper"; }
      { name = "fzf-vim"; }
    ];
  };
}
```

With this VIM the commands `:Files`, `:GFiles` and `:GFiles?` work fine.

**Comment by Ma27** [comment]
I wrongly assumed that you meant the list of VIM plugins (i.e. in `vim_configurable`) with dependencies. In fact you're right though, adding `fzf-vim` should be sufficient, anything else is fairly counter-intuitive. I'll rebase your commit against master. If this is sufficient I'd merge :)

**Comment by Ma27** [comment]
Hmm although the build itself works fine, the evaluation of a `vim_configurable` expression as pasted above fails with the following error:

```
error: attribute 'rtp' missing, at /home/ma27/Projects/nixpkgs/pkgs/misc/vim-plugins/vim-utils.nix:246:54
(use '--show-trace' to show detailed location information)
```

**Comment by timokau** [comment]
Makes sense, since `buildVimPlugin` isn't used and that adds the `rtp` attribute which is used by the vam and vim-plug implementations.

Maybe we can come up with a proper `propagatedBuildInputs` or something like that.

**Comment by timokau** [comment]
Thinking about it again: Shouldn't adding `fzfWrapper` to `fzf-vim`'s dependencies be pretty much equivalent to manually adding `fzfWrapper` to the list of vim plugins?

**Comment by timokau** [comment]
Actually testing it again now, it *does* work to simply add `fzfWrapper` to the `dependencies` of `fzf-vim`. I probably opened this PR before looking deeper into the plugin system and fixing the dependency management for the native plugin implementation (https://github.com/NixOS/nixpkgs/pull/52767).

Can you confirm that this also works for you @Ma27?

File: pkgs/misc/vim-plugins/overrides.nix
Comment (Ma27) [inline pkgs/misc/vim-plugins/overrides.nix]: This breaks evaluation with the following message (since `fzfWrapper` is in the same overlay):

```
error: undefined variable 'fzfWrapper' at /home/ma27/Projects/nixpkgs/pkgs/misc/vim-plugins/overrides.nix:180:34
(use '--show-trace' to show detailed location information)
```

**Review by Ma27 — COMMENTED** [review]
Unfortunately this doesn't work for me (with fixed evaluation), `:Files` still breaks in a vim configured by `vim_configurable` with `fzf-vim` loaded as plugin.

**Comment by timokau** [comment]
It works with packages but fails with vam. Adding `fzf-vim` to the dependencies generates this vam vimrc:

```vim
let g:nix_plugin_locations = {}
let g:nix_plugin_locations['fzf-vim'] = "/nix/store/fkp4swn162iq7pp309srkp6z0sgl2fk6-vimplugin-fzf-vim-2018-12-11/share/vim-plugins/fzf-vim"
let g:nix_plugin_locations['fzf'] = "/nix/store/a9qnz59i7xgjlh57kg5jybqmdizqd3pb-vimplugin-fzf-0.17.5/share/vim-plugins/fzf"

let g:nix_plugin_locations['vim-addon-manager'] = "/nix/store/1xh8pflfs94f0cmii7cfq5j11k6by2aw-vimplugin-vim-addon-manager-2018-07-27/share/vim-plugins/vim-addon-manager"
[...]
call add(l, {'name': 'fzf-vim'})
```

while adding it manually to the dictionaries generates this:

```
let g:nix_plugin_locations = {}
let g:nix_plugin_locations['fzf-vim'] = "/nix/store/fkp4swn162iq7pp309srkp6z0sgl2fk6-vimplugin-fzf-vim-2018-12-11/share/vim-plugins/fzf-vim"
let g:nix_plugin_locations['fzfWrapper'] = "/nix/store/a9qnz59i7xgjlh57kg5jybqmdizqd3pb-vimplugin-fzf-0.17.5/share/vim-plugins/fzf"
[...]
call add(l, {'name': 'fzf-vim'})
call add(l, {'name': 'fzfWrapper'})
```

So adding it to the deps uses the `pname`, while adding it manually uses the attribute name. Also, the `call` line isn't properly generated.

**Comment by timokau** [comment]
This is probably something I messed up in my refactor. I pushed a commit that works at least in this case. However with vam, I really don't know what I'm doing. I removed the "vam plugin -> toNix" thing completely and replaced it by a simple name dictionary. That is probably not valid in some cases.

**Comment by Ma27** [comment]
WIth the latest patches the `fzf-vim` problem is solved with vam as well. I also tested your change against my rather complex ViM setup, so there's a certain chance that your vam change is correct :)

Pinging @Mic92 as he worked quite much on the vim-plugins sub-package.

**Comment by Mic92** [comment]
I wonder if we should reduce the number of plugin manager we support?
Are there use cases for having all of them?

**Comment by Mic92** [comment]
Building and running all the vim tests seems to work:

```console
$ nix-build -A vimUtils
```

**Comment by Ma27** [comment]
> I wonder if we should reduce the number of plugin manager we support?

No idea admittedly. I've been using vam for ~2 years now and I never had a reason to change that.

File: pkgs/misc/vim-plugins/overrides.nix
Comment (Mic92) [inline pkgs/misc/vim-plugins/overrides.nix]: Do we need to expose it then to the user?

Comment (Mic92) [reply pkgs/misc/vim-plugins/overrides.nix]: Just having `fzf-vim` there would create less confusion what the right package is.

Comment (timokau) [reply pkgs/misc/vim-plugins/overrides.nix]: I think it makes sense to keep it because

1. Backwards compatibility. No need to break the configuration of some users who have added `fzfWrapper`.
2. According to [upstream's readme](https://github.com/junegunn/fzf.vim):

> fzf in itself is not a Vim plugin, and the official repository only provides the basic wrapper function for Vim and it's up to the users to write their own Vim commands with it

So that sounds like `fzfWrapper` is intended to be used by end-users. To be honest I doubt that there are many users that do this, but it makes sense to support.

**Comment by timokau** [comment]
> I wonder if we should reduce the number of plugin manager we support?
Are there use cases for having all of them?

With my refactors of the plugin infrastructure, I've noticed that pretty much every corner case is used by *somebody* (https://xkcd.com/1172/). While I wouldn't be opposed to remove stuff, there should be a proper deprecation period and I don't think we have deprecation quite figured out yet.

Also there would be some bikeshedding to be done to select the "one true" implementation. VAM and "vim packages" are probably the only serious contenders. I'd argue that vim packages are the way to go nowadays (native upstream support, provides everything we need), but I'm sure some disagree. Also most "legacy" configurations will be VAM.

File: pkgs/misc/vim-plugins/overrides.nix
Comment (Ma27) [reply pkgs/misc/vim-plugins/overrides.nix]: In that case I'd keep `fzfWrapper` for now (at least) and do a proper deprecation phase.

**Comment by Ma27** [comment]
> While I wouldn't be opposed to remove stuff, there should be a proper deprecation period and I don't think we have deprecation quite figured out yet.

Did you encounter any behavioral differences? Or is it possible that your changes in the plugin attr set breaks functionality?

> VAM and "vim packages" are probably the only serious contenders. I'd argue that vim packages are the way to go nowadays (native upstream support, provides everything we need), but I'm sure some disagree. Also most "legacy" configurations will be VAM.

If this simplifies things and the migration is not too hard, I'd say :+1: Unless I'm missing something, there shouldn't a big advantage from VAM.

Do you want to open an issue to discuss about this?

**Comment by timokau** [comment]
> > While I wouldn't be opposed to remove stuff, there should be a proper deprecation period and I don't think we have deprecation quite figured out yet.
> 
> Did you encounter any behavioral differences? Or is it possible that your changes in the plugin attr set breaks functionality?

Behavioral differences between what? I haven't tested the VAM version thoroughly (i.e. I just made sure it starts and `:Files` works). Its very possible that the change from `toNix p` to just the name breaks functionality. I don't know what that call is used for. It seemed like in practice, only the name ends up in there anyways. But I haven't done the research to confirm that.

I've used the old `symlinkJoin` version with vim packages since I've opened this PR, without any issues or behavioral differences.

> > VAM and "vim packages" are probably the only serious contenders. I'd argue that vim packages are the way to go nowadays (native upstream support, provides everything we need), but I'm sure some disagree. Also most "legacy" configurations will be VAM.
> 
> If this simplifies things and the migration is not too hard, I'd say +1 Unless I'm missing something, there shouldn't a big advantage from VAM.
> 
> Do you want to open an issue to discuss about this?

I've opened https://github.com/NixOS/nixpkgs/issues/56338. I do not intend to remove anything myself though, as I don't want to deal with the inevitable fallout of people complaining.

**Comment by Ma27** [comment]
> Behavioral differences between what? I haven't tested the VAM version thoroughly (i.e. I just made sure it starts and `:Files` works). Its very possible that the change from `toNix p` to just the name breaks functionality. I don't know what that call is used for. It seemed like in practice, only the name ends up in there anyways. But I haven't done the research to confirm that.

I wanted to know if you observed any differences (or breakage) after changing the `toNix` expression.

I can have a more detailed look at this at the end of the week, but I'm not sure about the consequences of this change as well ATM.

**Comment by lrworth** [comment]
Just dropping by. I haven't tried this branch, but in nixpkgs-unstable right now, it looks like there's a missing dependency on actual `fzf`, so `:Files:` prompts to download a binary. `nix-env -i fzf` fixes it.

**Review by Ma27 — COMMENTED** [review]
> Just dropping by. I haven't tried this branch, but in nixpkgs-unstable right now, it looks like there's a missing dependency on actual fzf, so :Files prompts to download a binary. nix-env -i fzf fixes it.

You're right, this probably got unnoticed, at least I have `fzf` installed on my system. @timokau I assume this could be fixed by simply patching the Vim script's sources?

Judging from the sources, I currently don't think that this actually breaks a user's setup, we simply add transitive dependencies (if not specified manually in `pluginDictionaries`), so if we hit an edge case, it's probably because of wrongly specified dependencies in `overrides.nix`.

I tested this with my historically-grown Vim config and didn't encounter further issues. Also I tested several plugins with dependencies (fzf-vim, vim-bazel, ensime-vim) and the patch seems fine.

I'd suggest we merge this for now (not sure if this should be backported) and wait for further feedback.

**Comment by zimbatm** [comment]
Is anything holding back this PR?

**Comment by Ma27** [comment]
Ouch, I totally forgot about that one, sorry :/

I fixed this locally and didn't encounter the problem since then. From my perspective this appears fine, @Mic92 @timokau opinions?

**Comment by zimbatm** [comment]
No worries, I just hit the same issue today so finding this PR was quite helpful actually.

What will happen to users who already have both plugins in their list, will it de-duplicate?

**Comment by Ma27** [comment]
I'm not entirely sure anymore as it's some time ago that I had a closer look at this. Let me check what happens tonight on my setup ok? :)

**Comment by timokau** [comment]
I don't remember the details with the different plugin managers, but if this works for everyone (or at least doesn't break anything), we can merge. Can't believe it has been 8 months already, didn't seem that long ago.

Personally I've been using the original `symlinkJoin` approach in my [vim packages](https://github.com/timokau/dotfiles/blob/ec689329c269ebde9c9e3b12fdf3021b62a39120/home/neovim.nix#L337) based configuration and kind of forgot about it.

**Review by Ma27 — APPROVED** [review]
As requested by @zimbatm I tested if this works with the old configuration (it appears to deduplicate) and with a vim config that only uses `fzf-vim`. As this is a pretty confusing bug and I guess that it took most of us to find out the cause (or this PR) I guess that it's a good thing to merge this now :+1:

**Comment by Ma27** [comment]
@timokau thanks!

**Pull request merged**
